pkg/providers/vsphere/validator.go

```diff
@@ -0,0 +1,292 @@
+		return errors.New("VSphereMachineConfig VM resourcePool for worker nodes is not set or is empty")
+	}
+
+	if vsphereClusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if vsphereClusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef == nil {
+			return errors.New("must specify machineGroupRef for etcd machines")
+		}
+		etcdMachineConfig = vsphereClusterSpec.etcdMachineConfig()
+		if etcdMachineConfig == nil {
+			return fmt.Errorf("cannot find VSphereMachineConfig %v for etcd machines", vsphereClusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name)
+		}
+		if len(etcdMachineConfig.Spec.Datastore) <= 0 {
+			return errors.New("VSphereMachineConfig datastore for etcd machines is not set or is empty")
+		}
+		if len(etcdMachineConfig.Spec.Folder) <= 0 {
+			logger.Info("VSphereMachineConfig folder for etcd machines is not set or is empty. Will default to root vSphere folder.")
+		}
+		if len(etcdMachineConfig.Spec.ResourcePool) <= 0 {
+			return errors.New("VSphereMachineConfig VM resourcePool for etcd machines is not set or is empty")
+		}
+	}
+
+	if err := p.validateControlPlaneIp(vsphereClusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host); err != nil {
+		return err
+	}
+
+	if err := p.govc.ValidateVCenterSetup(ctx, vsphereClusterSpec.datacenterConfig, &p.selfSigned); err != nil { // TODO: remove side effects from this implementation or directly move it to set defaults (try not to pass a pointer to bool to update its value, it's very difficult to follow)
+		return fmt.Errorf("error validating vCenter setup: %v", err)
+	}
+
+	for _, config := range vsphereClusterSpec.machineConfigsLookup {
+		err := p.govc.ValidateVCenterSetupMachineConfig(ctx, vsphereClusterSpec.datacenterConfig, config, &p.selfSigned) // TODO: remove side effects from this implementation or directly move it to set defaults (pointer to bool is not needed)
+		if err != nil {
+			return fmt.Errorf("error validating vCenter setup for VSphereMachineConfig %v: %v", config.Name, err)
+		}
+	}
+
+	if controlPlaneMachineConfig.Spec.OSFamily != workerNodeGroupMachineConfig.Spec.OSFamily {
+		return errors.New("control plane and worker nodes must have the same osFamily specified")
+	}
+
+	if etcdMachineConfig != nil && controlPlaneMachineConfig.Spec.OSFamily != etcdMachineConfig.Spec.OSFamily {
+		return errors.New("control plane and etcd machines must have the same osFamily specified")
+	}
+
+	if err := p.validateSSHUsername(controlPlaneMachineConfig); err == nil {
+		if err = p.validateSSHUsername(workerNodeGroupMachineConfig); err != nil {
+			return fmt.Errorf("error validating SSHUsername for worker node VSphereMachineConfig %v: %v", workerNodeGroupMachineConfig.Name, err)
+		}
+		if etcdMachineConfig != nil {
+			if err = p.validateSSHUsername(etcdMachineConfig); err != nil {
+				return fmt.Errorf("error validating SSHUsername for etcd VSphereMachineConfig %v: %v", etcdMachineConfig.Name, err)
+			}
+		}
```

Comment:
why are these nested inside controlplaneMachine's sshUsername check?

Reply:
Tbh, I'm not sure, this is just copy paste
I was planning on refactoring this whole method separately, including drying out the machine config validation, since it's the same for all machine configs